### PR TITLE
Add new opcodes

### PR DESCRIPTION
**Pull request by hkalodner**
This PR adds a number of new opcodes
- SETGAS
- PUSHGAS
- XGET
- XSET

### CODE REVIEW

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 103
- Added 228
           

Complexity increasing per file
==============================
- packages/arb-validator/challenges/executionChallenger.go  10
- packages/arb-validator/rollup/preparedAssertion.go  1
- packages/arb-validator-core/ethbridgetest/protocoltester/protocoltester.go  3
- packages/arb-avm-cpp/gotest/gotest.go  2
- packages/arb-validator/challenges/messagesDefender.go  12
- packages/arb-util/value/codePointStub.go  3
- packages/arb-validator-core/ethbridgetest/protocol_test.go  3
- packages/arb-validator/proofmachine/proofmachine.go  1
- packages/arb-util/value/values.go  1
- packages/arb-validator/challenges/inboxTopDefender.go  9
- packages/arb-validator/rollup/rollupEthBridge_test.go  5
- packages/arb-compiler-evm/arbitrum/marshall.py  1
- packages/arb-validator-core/ethbridgetest/onestepprooftester/onestepprooftester.go  3
- packages/arb-validator/challenges/challengeUtils.go  3
- packages/arb-validator-core/ethbridgetest/main_test.go  4
- packages/arb-validator/challenges/messagesChallenger.go  10
- packages/arb-validator/challenges/inboxTopChallenger.go  9
- packages/arb-validator/challenges/messagesChallenger_test.go  4
- packages/arb-validator/challenges/executionDefender.go  10
- packages/arb-validator/proofmachine/machine_test.go  5
- packages/arb-util/value/hashPreImage.go  3
- packages/arb-validator-core/utils/rpc.go  4
- packages/arb-validator/proofmachine/machinetester/machinetester.go  3
         

Complexity decreasing per file
==============================
+ packages/arb-validator/loader/loader.go  -2
+ packages/arb-validator-core/ethbridgetest/challengetester/challengetester.go  -7
+ packages/arb-validator/challenges/inboxTopChallenge_test.go  -1
+ packages/arb-util/value/tuple.go  -1
+ packages/arb-validator-core/ethbridgetest/rolluptester/rolluptester.go  -8
+ packages/arb-validator/challenges/testHelper.go  -8
+ packages/arb-validator/challenges/executionChallenge_test.go  -1
         

Clones added
============
- packages/arb-validator-core/ethbridge/messagesChallenge.go  2
- packages/arb-validator/challenges/executionChallenger.go  4
- packages/arb-validator-core/ethbridgetest/valuetester/valuetester.go  6
- packages/arb-validator/cmdhelper/cmdhelper.go  1
- packages/arb-validator-core/ethbridgetest/protocoltester/protocoltester.go  14
- packages/arb-validator/challenges/messagesDefender.go  6
- packages/arb-validator/challenges/inboxTopDefender.go  2
- packages/arb-validator/rollup/rollupEthBridge_test.go  6
- packages/arb-validator-core/arbbridge/arbRollup.go  1
- packages/arb-validator-core/ethbridgetest/onestepprooftester/onestepprooftester.go  13
- packages/arb-validator/challenges/challengeUtils.go  2
- packages/arb-validator-core/ethbridgetest/main_test.go  2
- packages/arb-validator/challenges/messagesChallenger.go  6
- packages/arb-validator/challenges/inboxTopChallenger.go  3
- packages/arb-validator/challenges/messagesChallenger_test.go  1
- packages/arb-validator/challenges/executionDefender.go  4
- packages/arb-validator-core/ethbridgetest/value_test.go  2
- packages/arb-validator-core/ethbridgetest/rolluptester/rolluptester.go  29
- packages/arb-validator/proofmachine/machinetester/machinetester.go  16
- packages/arb-validator/challenges/executionChallenge_test.go  2
- packages/arb-validator/structures/node_test.go  1
         

Clones removed
==============
+ packages/arb-util/value/codePoint.go  -2
+ packages/arb-validator-core/ethbridge/arbRollup.go  -1
+ packages/arb-validator/proofmachine/proofmachine.go  -1
+ packages/arb-validator-core/arbbridge/arbClient.go  -2
+ packages/arb-util/value/values.go  -2
+ packages/arb-validator-core/ethbridgetest/challengetester/challengetester.go  -8
+ tests/fibgo/connection_test.go  -1
+ tests/fibgo/fibabi.go  -5
+ packages/arb-avm-cpp/tests/checkpoint.cpp  -2
+ packages/arb-avm-cpp/utils/bigint_utils.hpp  -1
+ packages/arb-validator/challenges/testHelper.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5783203&bid=18933395)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- packages/arb-validator/proofmachine/proofmachine.go  1
         

Clones removed
==============
+ packages/arb-avm-cpp/tests/checkpoint.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5783203&bid=18933395)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- packages/arb-validator/proofmachine/proofmachine.go  1
         

Clones removed
==============
+ packages/arb-avm-cpp/tests/checkpoint.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5783203&bid=18933395)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- packages/arb-validator/proofmachine/proofmachine.go  1
         

Clones removed
==============
+ packages/arb-avm-cpp/tests/checkpoint.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5783203&bid=18933395)

File: packages/arb-bridge-eth/contracts/arch/OneStepProof.sol [hunk 1]
```diff
@@ -812,6 +850,26 @@ library OneStepProof {
         return true;
     }
 
+    function executeSetgasInsn(
```
Comment (minhptruong) [inline packages/arb-bridge-eth/contracts/arch/OneStepProof.sol:853]: capital G

File: packages/arb-bridge-eth/contracts/arch/OneStepProof.sol [hunk 2]
```diff
@@ -1253,7 +1331,10 @@ library OneStepProof {
                 (!_data.didInboxInsn && opCode != OP_INBOX),
             "Invalid didInboxInsn claim"
         );
-        if (opCode == OP_ADD) {
+        if (startMachine.arbGasRemaining < opGasCost(opCode)) {
+            endMachine.arbGasRemaining = ((1 << 128) + 1) * ((1 << 128) - 1); // = MaxUint256
```
Comment (minhptruong) [inline packages/arb-bridge-eth/contracts/arch/OneStepProof.sol:1335]: is this better thank having a constant?

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- packages/arb-validator/proofmachine/proofmachine.go  1
         

Clones removed
==============
+ packages/arb-avm-cpp/tests/checkpoint.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5783203&bid=18933395)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- packages/arb-validator/proofmachine/proofmachine.go  1
         

Clones removed
==============
+ packages/arb-avm-cpp/tests/checkpoint.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5783203&bid=18933395)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- packages/arb-validator/proofmachine/proofmachine.go  1
         

Clones added
============
- packages/arb-avm-cpp/tests/opcodes.cpp  2
         

Clones removed
==============
+ packages/arb-avm-cpp/tests/checkpoint.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5783203&bid=18933395)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- packages/arb-validator/proofmachine/proofmachine.go  1
         

Clones added
============
- packages/arb-avm-cpp/tests/opcodes.cpp  2
         

Clones removed
==============
+ packages/arb-avm-cpp/tests/checkpoint.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5783203&bid=18933395)

File: packages/arb-avm-cpp/avm/src/machinestate/machinestate.cpp [hunk 3]
```diff
@@ -208,6 +247,16 @@ BlockReason MachineState::runOne() {
...
+    } else if (arb_gas_remaining <
+               InstructionArbGasCost.at(instruction.op.opcode)) {
+        arb_gas_remaining = max_arb_gas_remaining;
+        state = Status::Error;
+        context.numSteps++;
+        if (!errpc.is_err) {
+            pc = errpc;
+            state = Status::Extensive;
+        }
+        return NotBlocked();
```
[the remark this answers is not in the document]
Comment (hkalodner) [reply packages/arb-avm-cpp/avm/src/machinestate/machinestate.cpp]: This was correct, but I've cleaned up and clarified the code

File: packages/arb-bridge-eth/contracts/arch/OneStepProof.sol (hunk 2 above)
Comment (hkalodner) [reply packages/arb-bridge-eth/contracts/arch/OneStepProof.sol]: Changed to using a constant

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- packages/arb-validator/proofmachine/proofmachine.go  1
         

Clones added
============
- packages/arb-avm-cpp/tests/opcodes.cpp  2
         

Clones removed
==============
+ packages/arb-avm-cpp/tests/checkpoint.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5783203&bid=18933395)

**Comment by hkalodner** [comment]
I had to manually fix a merge conflict so this will need review again

**Pull request merged**